### PR TITLE
Added an optional timeout in the endpoint configuration of remote services

### LINKED ISSUE #1277: Add an optional timeout in the endpoint configuration of remote services
### The problem faced currently?

Right now, we can provide a `timeout` while calling an endpoint of a remote service. However, this means we need to change the client-side code that is consuming the remote service in order to change this timeout.  

### How can we solve it?

Let's add the `timeout` field (time in seconds) in the configuration of each endpoint in remote service. This field should be optional and space cloud should take a default value of `10` if Mission Control doesn't provide a value for `timeout`.

This `timeout` field can go in the advanced configuration panel of the remote endpoint configuration form.  

>If you want this feature to be implemented, give it a thumbs up reaction, so that we can determine which features are important to you.
>👍

### PR DESCRIPTION
**Pull request by KavishShah09**
Fixes #1277 for area/gateway

### CODE REVIEW

File: gateway/managers/syncman/syncman_services.go [hunk 1]
```diff
@@ -26,6 +27,14 @@ func (s *Manager) SetService(ctx context.Context, project, service string, value
 	if projectConfig.Modules.Services.Services == nil {
 		projectConfig.Modules.Services.Services = config.Services{}
 	}
+
+	// Check timeout field and add default value if not present
+	for _, val := range value.Endpoints {
+		if val.Timeout == 0 {
+			val.Timeout = int(10 * time.Second)
```
Comment (sharadregoti) [inline gateway/managers/syncman/syncman_services.go:34]:
```suggestion
			val.Timeout = 10
```

**Review by sharadregoti — CHANGES_REQUESTED** [review]

**Review by sharadregoti — CHANGES_REQUESTED** [review]

**Review by sharadregoti — CHANGES_REQUESTED** [review]

**Review by sharadregoti — APPROVED** [review]

File: gateway/modules/functions/functions.go [hunk 2]
```diff
@@ -98,3 +102,14 @@ func (m *Module) SetConfig(project string, c *config.ServicesModule) error {
 	}
 	return nil
 }
+
+// GetEndpointContextTimeout returns the endpoint timeout of particular remote-service
+func (m *Module) GetEndpointContextTimeout(ctx context.Context, service, function string) (int, error) {
+
```
Comment (YourTechBud) [inline gateway/modules/functions/functions.go:108]: Acquire a read lock here

File: gateway/modules/functions/functions.go [hunk 3]
```diff
@@ -98,3 +102,14 @@ func (m *Module) SetConfig(project string, c *config.ServicesModule) error {
 	}
 	return nil
 }
+
+// GetEndpointContextTimeout returns the endpoint timeout of particular remote-service
+func (m *Module) GetEndpointContextTimeout(ctx context.Context, service, function string) (int, error) {
```
Comment (YourTechBud) [inline gateway/modules/functions/functions.go:107]: Shift this function to `operations.go`

File: gateway/server/handlers/functions.go [hunk 4]
```diff
@@ -36,24 +36,26 @@ func HandleFunctionCall(modules *modules.Modules) http.HandlerFunc {
 		// Get the JWT token from header
 		token := utils.GetTokenFromHeader(r)
 
-		// Set a default timeout value
-		if req.Timeout == 0 {
-			req.Timeout = 10 // set default context to 10 second
+		timeOut, err := functions.GetEndpointContextTimeout(r.Context(), serviceID, function)
+		if err != nil {
+			_ = helpers.Logger.LogError(helpers.GetRequestID(r.Context()), fmt.Sprintf("Could not find endpoint (%s) for service (%s)", serviceID, function), err, nil)
```
Comment (YourTechBud) [inline gateway/server/handlers/functions.go:41]: No need to log again here

File: gateway/server/handlers/functions.go [hunk 5]
```diff
@@ -36,24 +36,26 @@ func HandleFunctionCall(modules *modules.Modules) http.HandlerFunc {
...
 
-		actions, reqParams, err := auth.IsFuncCallAuthorised(ctx, projectID, service, function, token, req.Params)
+		actions, reqParams, err := auth.IsFuncCallAuthorised(ctx, projectID, serviceID, function, token, req.Params)
 		if err != nil {
 			_ = helpers.Response.SendErrorResponse(ctx, w, http.StatusForbidden, err.Error())
 			return
 		}
 
-		status, result, err := functions.CallWithContext(ctx, service, function, token, reqParams, req.Params)
+		status, result, err := functions.CallWithContext(ctx, serviceID, function, token, reqParams, req.Params)
```
Comment (YourTechBud) [inline gateway/server/handlers/functions.go:56]: Call the `utils.ExtractRequestParams` before this line

**Review by YourTechBud — CHANGES_REQUESTED** [review]

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**